### PR TITLE
[WIP] Massive refactor

### PR DESCRIPTION
**Pull request by JakeGinnivan**
Getting some early visibility of this refactoring. It is now compiling and all the tests are switched to new API.

Will comments on some to callout the major changes. As I get tests passing things will likely change more.

Also, INotifyPropertyChanged support was dropped, will reintroduce once this is working

### CODE REVIEW

File: src/ReactGraph/DependencyEngine.cs [hunk 1]
```diff
@@ -117,5 +98,272 @@ public void CheckCycles()
 
             throw new CycleDetectedException(sb.ToString().Trim());
         }
+
+        public void AddExpression<T>(ISourceDefinition<T> source, ITargetDefinition<T> target, Action<Exception> onError)
```
Comment (JakeGinnivan) [inline src/ReactGraph/DependencyEngine.cs:133]: This is now the way nodes are added to the graph. Always in pairs. 

Once is a source the other a target, and an error handler has to be provided.

**Comment by JakeGinnivan** [comment]
@odeheurles this is getting closer now.

One of the biggest side effects is that roots are now part of the graph. i.e `viewModel.Foo` will have two nodes in the graph.

I was thinking we could tag then and they could be excluded from any visualisations.

File: src/ReactGraph.Tests/ExpressionParserTest.cs [hunk 2]
```diff
@@ -54,19 +47,18 @@ public void GetInvertedHarderNode()
...
             viewModel.RegeneratePaymentSchedule(false);
             Expression<Func<bool>> expr = () => !viewModel.PaymentSchedule.HasValidationError;
-            var node = expressionParser.GetFormulaInfo(expr);
-            node.Dependencies.Count.ShouldBe(1);
-            var validationErrorNode = node.Dependencies[0];
-            validationErrorNode.RootInstance.ShouldBeSameAs(viewModel);
-            validationErrorNode.Dependencies.Count.ShouldBe(1);
-            validationErrorNode.ShouldBeOfType<MemberDependencyDescriptor<bool>>()
-                .MemberInfo.ShouldBe(typeof(ScheduleViewModel).GetProperty("HasValidationError"));
+            var node = ExpressionParser.GetChildSources(expr);
```
Comment (odeheurles) [inline src/ReactGraph.Tests/ExpressionParserTest.cs:50]: rename the node variable?

File: src/ReactGraph.Tests/ExpressionParserTest.cs [hunk 3]
```diff
@@ -74,8 +66,38 @@ public void SimpleMethod()
         {
             var simple = new SimpleWithNotification();
             Expression<Func<int>> expr = () => Negate(simple.Value);
-            var node = expressionParser.GetFormulaInfo(expr);
-            node.Dependencies.Count.ShouldBe(1);
+            var node = ExpressionParser.GetChildSources(expr);
```
Comment (odeheurles) [inline src/ReactGraph.Tests/ExpressionParserTest.cs:69]: same here?

File: src/ReactGraph/BuilderBase.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            var memberExpression = expression.Body as MemberExpression;
+            if (memberExpression == null) return false;
+            var propertyInfo = memberExpression.Member as PropertyInfo;
+            var fieldInfo = memberExpression.Member as FieldInfo;
+
+            if (propertyInfo == null && fieldInfo == null) return false;
+            if (propertyInfo != null && !propertyInfo.CanWrite) return false;
+            if (fieldInfo != null && fieldInfo.IsInitOnly) return false;
+
+            return true;
```
Comment (odeheurles) [inline src/ReactGraph/BuilderBase.cs:22]: shouldn't this return false by default and true only when it's actually writable?

File: src/ReactGraph/Construction/ExpressionStringBuilder.cs [hunk 5]
```diff
@@ -22,14 +22,20 @@ public static string ToString(Expression expression)
...
+                return s.Substring(1, s.Length - 2);
+            return s;
         }
 
         protected override Expression VisitLambda<T>(Expression<T> node)
         {
-            Out("(");
-            Out(String.Join(",", node.Parameters.Select(n => n.Name)));
-            Out(") => ");
+            if (node.Parameters.Any())
```
Comment (odeheurles) [inline src/ReactGraph/Construction/ExpressionStringBuilder.cs:33]: better

File: src/ReactGraph/Construction/ExpressionStringBuilder.cs [hunk 6]
```diff
@@ -22,14 +22,20 @@ public static string ToString(Expression expression)
         {
             var visitor = new ExpressionStringBuilder();
             visitor.Visit(expression);
-            return visitor.builder.ToString();
+            var s = visitor.builder.ToString();
+            if (s.StartsWith("(") && s.EndsWith(")"))
```
Comment (odeheurles) [inline src/ReactGraph/Construction/ExpressionStringBuilder.cs]: are you sure?

"(A + B) \* (C - D)" will print "A + B) \* (C - D"
?

Comment (JakeGinnivan) [inline src/ReactGraph/Construction/ExpressionStringBuilder.cs]: Good point, damn

**Comment by JakeGinnivan** [comment]
Root nodes are now in the graph. Currently this makes visualisations go from:

image

to

image

Is this an issue, or should I fix before merge

**Comment by JakeGinnivan** [comment]
Rebased, removed the root nodes from visualisations and fixed your comment about the brackets.

Notify property changed not done yet, will see if I can get it done in the morning

**Comment by JakeGinnivan** [comment]
I've fixed this :P

Sent from my Windows Phone

**Comment by JakeGinnivan** [comment]
Notify property changed support added. It is a hack, but works. Needs refactoring

**Pull request merged**